### PR TITLE
[第七組] U4W2 Refactor/error handling

### PR DESCRIPTION
**Pull request by haruuu1214**
## Summary
This PR includes updates to various modules and adds error handling to improve the robustness and functionality of the application. The following changes are included:

### `src/auth/models.py`
- Improved authentication models to enhance security and functionality.

### `src/config.py`
- Updated configuration settings to support new features and improve flexibility.

### `src/main.py`
- Enhanced with additional startup and shutdown events to ensure proper initialization and cleanup.

### `src/news/router.py`
- Added error handling and logging to improve the reliability of news-related endpoints.

### `src/news/service.py`
- Improved service logic to handle edge cases and improve performance.

### `src/pricing/router.py`
- Updated with new pricing endpoints to support additional pricing features.

### `src/users/router.py`
- Enhanced user registration and added error handling to provide better feedback and reliability.

### `tests/integration/test_news_endpoint.py`
- Added integration tests for news endpoints to ensure correctness and reliability.

## Testing
- Unit tests and integration tests have been added and updated to cover the new changes.
- Mocked responses and error scenarios to ensure comprehensive test coverage.

## Notes
- Ensure that the new configuration settings are properly set in the environment variables before deploying.
- Additional tests may be required to cover all edge cases and error handling scenarios.

Please review the changes and provide feedback.

### CODE REVIEW

**Comment by Steven0811** [comment]
對於logger跟各種exception的定義使用了Template Method Pattern，並且在定義base時使用了很多抽象方法，對於OOP的概念有良好的展現
但是你雖然定義了各種可能會發生的exception，但是在應用上有一些問題
```python
# users\router.py
except APIException as e:
    logger.error(f"User login failed: {e}")
    raise e
except Exception as e:
    db_session.rollback()
    logger.error(f"User login failed: {e}")
    raise InternalServerErrorException(e)
```
這裡直接使用APIException作為第一項exception的判定，但是你在APIException底下明明還有很多繼承他的子類，應該要明確指出是哪exception而非直接使用作為base的APIException
```python
# news\router.py
try:
    detailed_news_info = convert_news_to_dict(process_news_item(news_item))
    detailed_news_info["id"] = next(article_id_counter)
    extracted_news_list.append(detailed_news_info)
    logger.info(f"Search news success: {detailed_news_info['title']}")
except Exception as error:
    logger.error(f"Search news failed: {news_item.url}", error=error)
```
還有這裡直接使用Exception，照理來說你應該要使用你有定義的各種exception，未知的錯誤可用InternalServerErrorException

**Comment by mlcp29th-hi** [comment]
1. **Consistent Custom Exceptions**:
   - It appears that you have created a well-defined set of custom exceptions in the `exceptions_handler` package. This structure is beneficial for clarity and debugging. Ensure that you consistently use these exceptions throughout your code for all error scenarios, not just in critical areas, to maintain uniformity and make it easier to track issues.

2. **Specificity of Exceptions**:
   - Instead of a generic `InternalServerErrorException`, try to throw more specific exceptions that could provide clearer insights into what went wrong during execution. For example, instead of just capturing an exception in the `fetch_news_with_details` function, consider whether you can provide more context regarding the failure. If it's a database-related issue, a specific `DatabaseConnectionException` (which you could define) might be more suitable.

3. **Logging Enhancements**:
   - Make sure to log not only the errors but also information that could be useful in understanding the context, such as the parameters passed to functions when errors occur. This will greatly enhance your ability to perform post-mortem debugging.

4. **Fail-Safe Mechanisms**:
   - When dealing with external calls (like fetching news articles), implement retry mechanisms or fallback strategies when exceptions are caught. This will help prevent a single point of failure affecting the entire functionality.

5. **Exception Handling in Async Contexts**:
   - In your async function `search_news`, ensure that the error handling strategy (logging and raising exceptions) is consistent, especially since it interacts with external services (e.g., OpenAI's API). You might want to create specific response codes and messages for various failure scenarios to aid client-side error handling.

6. **Graceful Shutdown**:
   - In your shutdown handler (`shutdown_scheduler`), you should handle exceptions similarly to how you handle startup errors. Log the error, if any, and ensure that the service can gracefully close regardless of exceptions, which is important for resource management.

7. **Testing of Error Scenarios**:
   - Ensure that your tests cover error scenarios as well. For example, simulate API failures in your tests to verify that the system handles exceptions correctly without crashing.

8. **Input Validation**:
   - While you are already validating the size of the username and password in your `login_for_access_token` function, consider whether there are other forms of validation that could provide additional safeguards, such as checking for disallowed characters or SQL injection protections.

By enhancing your error handling practices, you'll not only improve the robustness of your application but also create a more maintainable and easily debugged codebase. Please consider these suggestions as you continue to refine your implementation!

**Comment by ptchen314** [comment]
## 阿怎麼每個人都在修CI

改的很好，try expecr的部分有詳細實作，不是只有Exception as e。
PR寫的非常詳細，我十分讚賞。

**Comment by jxuanyang** [comment]
這個 PR 提升應用穩定性與功能性， error handling增加的每個功能很明確

**Comment by Yi-Cheng-Wang** [comment]
## Advantages

1. **Unified Error-Handling Structure**:  
   Using custom exception classes (e.g., `APIException` and its subclasses) allows for consistent handling of different types of errors across the application, making the codebase easier to maintain and extend.  

2. **Detailed Error Information**:  
   The error-handling logic includes detailed error information (e.g., `exc_info=True`), which helps log stack traces, enabling developers to debug issues more effectively.  

3. **HTTP Status Code Usage**:  
   Returning appropriate HTTP status codes (e.g., 401, 404, 500) based on the error type adheres to RESTful API design principles and helps clients better understand the nature of the error.  

4. **Logging**:  
   Using logging mechanisms (e.g., `logger.error`) during error handling helps track issues in production environments and provides runtime visibility into application behavior.  

## Disadvantages

1. **Over-Reliance on Custom Exceptions**:  
   While custom exceptions offer flexibility, overusing them can make the codebase overly complex, especially when handling multiple exception types, requiring numerous exception classes to be written.  

2. **Lack of Specific Error-Handling Logic**:  
   In some cases, error handling may simply involve logging the error and re-raising the exception, which might leave users without actionable information or solutions to the problem.  

3. **Redundancy in Error Handling**:  
   Error-handling logic may appear repeatedly in some functions, leading to code duplication and increased maintenance costs.  

4. **Lack of User-Friendly Error Messages**:  
   While the backend logs detailed error information, the messages returned to users may be overly technical, making it difficult for non-technical users to understand the issue.  

## Summary

The error-handling mechanism in this code demonstrates strong consistency and maintainability but could be improved in terms of specific error-handling logic and user-friendliness. Moving forward, it’s recommended to strike a balance between flexibility and simplicity in error handling and to provide more specific and user-friendly error messages to end users.

**Review by samlu00412 — COMMENTED** [review]
對於error handling的分工非常分明，整個project很有OOP的概念，但因為有點過於瑣碎，所以擴充後的可維護性沒有想像中的高。但整體上還是非常好的

**Comment by Diego0923** [comment]
錯誤處理結構清晰，展現良好OOP概念，
但建議避免過度依賴基類異常，改用具體異常提升可讀性，
統一使用自定義異常類，減少重複邏輯並提升維護性。
這份PR的改動值得肯定

File: .github/workflows/ci.yaml
Comment (dee9c-git) [inline .github/workflows/ci.yaml:1]: I would think openai_key or openai_token will be a bit more precise.

File: backend/src/users/router.py [hunk 1]
```diff
@@ -34,16 +41,31 @@ async def login_for_access_token(
...
+        user = check_user_password_is_correct(db_session, user_data.username, user_data.password)
 
-    # 創建訪問令牌，30 分鐘有效期
-    access_token = create_access_token(
-        data={"sub": str(user.username)},
-        expires_delta=timedelta(minutes=Config.News.NEWS_FETCH_INTERVAL_TIME)
-    )
-    
-    return {"access_token": access_token, "token_type": "bearer"}
+        # 創建訪問令牌，30 分鐘有效期
```
Comment (dee9c-git) [inline backend/src/users/router.py:55]: Has this changed to 100?

File: backend/src/users/router.py [hunk 2]
```diff
@@ -34,16 +41,31 @@ async def login_for_access_token(
...
     """
-    # 驗證使用者的帳號和密碼
-    user = check_user_password_is_correct(db_session, form_data.username, form_data.password)
+    try:
+        if len(user_data.username) > Config.Auth.MAX_USERNAME_SIZE:
+            logger.error(f"User login failed: {InvalidUsernameSizeException(len(user_data.username))}")
+            raise InvalidUsernameSizeException(len(user_data.username))
+        if len(user_data.password) > Config.Auth.MAX_PASSWORD_SIZE:
+            logger.error(f"User login failed: {InvalidPasswordSizeException(len(user_data.username))}")
+            raise InvalidPasswordSizeException(len(user_data.password))
```
Comment (dee9c-git) [inline backend/src/users/router.py:50]: This feels quite similar to auth/models.py:
```python
    @staticmethod
    def validate_username(username: str):
        if len(username) > Config.Auth.MAX_USERNAME_SIZE:
            raise ValueError("Username is too long")

    @staticmethod
    def validate_password(password: str):
        if len(password) > Config.Auth.MAX_PASSWORD_SIZE:
            raise ValueError("Password is too long")
```
I feel like there is redundant code here

File: backend/src/users/router.py [hunk 3]
```diff
@@ -53,10 +75,21 @@ def create_user(user_data: UserAuthSchema , db_session: Session = Depends(sessio
...
+        
+        hashed_password = pwd_context.hash(user_data.password)
+        new_user = User(username=user_data.username, hashed_password=hashed_password)
+        
+        db_session.add(new_user)
+        db_session.commit()
+        db_session.refresh(new_user)
```
Comment (dee9c-git) [inline backend/src/users/router.py:91]: Large ```try``` statement

**Review by dee9c-git — COMMENTED** [review]
I may have to look deeper to understand, but it seems a bit odd to have a mega exception_handler made up by a few smaller files. This makes it easy to use any made exception, but it can feel like you're swimming in exceptions to find the right one.
A few other things:

**Comment by hamster0717** [comment]
### **Strengths**
1. **Improved Error Handling:**
   - Introducing specific exceptions like `APIException`, `InternalServerErrorException`, and `NoResourceFoundException` enhances code clarity and error traceability.
   - Logging failures with `exc_info=True` provides more context for debugging.

2. **Enhanced Logging:**
   - Meaningful log messages (`logger.info`, `logger.error`) improve observability and traceability.
   - Adding logging to critical operations (e.g., fetching articles, handling upvotes) ensures better monitoring.
